test/statistical_testing/graph_testing.cpp

```diff
@@ -0,0 +1,109 @@
+#include <fstream>
+#include <iostream>
+#include "../../include/graph.h"
+#include "../util/graph_gen.h"
+#include "../util/graph_verifier.h"
+
+static inline int do_run() {
+    ifstream in{"./sample.txt"};
+    node_t n, m;
+    in >> n >> m;
+    Graph g{n};
+    int type, a, b;
+    while (m--) {
+      in >> type >> a >> b;
+      if (type == INSERT) {
+        g.update({{a, b}, INSERT});
+      } else g.update({{a, b}, DELETE});
+    }
+    g.set_cumul_in("./cumul_sample.txt");
+    try {
+        g.connected_components();
+    } catch (std::exception const &err) {
+        return 1;
+    }
+    return 0;
+}
+
+int small_graph_test(int runs) {
+    int failures = 0;
+    for (int i = 0; i < runs; i++) {
+        generate_stream({1024,0.002,0.5,0,"./sample.txt","./cumul_sample.txt"});
+        failures += do_run();
+    }
+    return failures;
+}
+
+int medium_graph_test(int runs) {
+    int failures = 0;
+    for (int i = 0; i < runs; i++) {
+        generate_stream({2048,0.002,0.5,0,"./sample.txt","./cumul_sample.txt"});
+        failures += do_run();
+    }
+    return failures;
+}
+
+int medium_with_iso_test(int runs) {
+    int failures = 0;
+    for (int i = 0; i < runs; i++) {
+        generate_stream({2048,0.002,0.5,0,"./sample.txt","./cumul_sample.txt"});
+        std::fstream graph_file{"./sample.txt"};
+        std::fstream cumul_file{"./cumul_sample.txt"};
+        graph_file.write("2070", 4); // increase the node size to create iso nodes
```

Comment:
what are iso nodes?

Reply:
So this is basically a test which approximates our kron streams where there are a few nodes which are guaranteed to not have any neighbors.